fabric-nix/my_project/wallpaper/TODO_SETTINGS.md
Usage: swww img [OPTIONS] <PATH>

Arguments:
  <PATH>
          Path to the image to display

Options:
  -o, --outputs <OUTPUTS>
          Comma separated list of outputs to display the image at.
          
          If it isn't set, the image is displayed on all outputs.
          
          [default: ]

      --resize <RESIZE>
          Whether to resize the image and the method by which to resize it
          
          [default: crop]

          Possible values:
          - no:   Do not resize the image
          - crop: Resize the image to fill the whole screen, cropping out parts that don't fit
          - fit:  Resize the image to fit inside the screen, preserving the original aspect ratio

      --fill-color <FILL_COLOR>
          Which color to fill the padding with when output image does not fill screen
          
          [default: 000000]

  -f, --filter <FILTER>
          Filter to use when scaling images (run swww img --help to see options).
          
          Available options are:
          
          Nearest | Bilinear | CatmullRom | Mitchell | Lanczos3
          
          These are offered by the fast_image_resize crate (https://docs.rs/fast_image_resize/2.5.0/fast_image_resize/). 'Nearest' is what I recommend for pixel art stuff, and ONLY for
          pixel art stuff. It is also the fastest filter.
          
          For non pixel art stuff, I would usually recommend one of the last three, though some experimentation will be necessary to see which one you like best. Also note they are all
          slower than Nearest.
          
          [default: Lanczos3]

  -t, --transition-type <TRANSITION_TYPE>
          Sets the type of transition. Default is 'simple', that fades into the new image
          
          Possible transitions are:
          
          none | simple | fade | left | right | top | bottom | wipe | wave | grow | center | any | outer | random
          
          The 'left', 'right', 'top' and 'bottom' options make the transition happen from that position to its opposite in the screen.
          
          'none' is an alias to 'simple' that also sets the 'transition-step' to 255. This has the effect of the transition finishing instantly
          
          'fade' is similar to 'simple' but the fade is controlled through the --transition-bezier flag
          
          'wipe' is similar to 'left' but allows you to specify the angle for transition with the `--transition-angle` flag.
          
          'wave' is similar to 'wipe' sweeping line is wavy
          
          'grow' causes a growing circle to transition across the screen and allows changing the circle's center position with the `--transition-pos` flag.
          
          'center' is an alias to 'grow' with position set to center of screen.
          
          'any' is an alias to 'grow' with position set to a random point on screen.
          
          'outer' is the same as grow but the circle shrinks instead of growing.
          
          Finally, 'random' will select a transition effect at random
          
          [env: SWWW_TRANSITION=]
          [default: simple]

      --transition-step <TRANSITION_STEP>
          How fast the transition approaches the new image.
          
          The transition logic works by adding or subtracting from the current rgb values until the old image transforms in the new one. This controls by how much we add or subtract.
          
          Larger values will make the transition faster, but more abrupt. A value of 255 will always switch to the new image immediately.
          
          This defaults to 2 when transition-type is 'simple', and 90 otherwise
          
          [env: SWWW_TRANSITION_STEP=]
          [default: 90]

      --transition-duration <TRANSITION_DURATION>
          How long the transition takes to complete in seconds.
          
          Note that this doesn't work with the 'simple' transition
          
          [env: SWWW_TRANSITION_DURATION=]
          [default: 3]

      --transition-fps <TRANSITION_FPS>
          Frame rate for the transition effect.
          
          Note there is no point in setting this to a value smaller than what your monitor supports.
          
          Also note this is **different** from the transition-step. That one controls by how much we approach the new image every frame.
          
          [env: SWWW_TRANSITION_FPS=]
          [default: 30]

      --transition-angle <TRANSITION_ANGLE>
          This is used for the 'wipe' and 'wave' transitions. It controls the angle of the wipe
          
          Note that the angle is in degrees, where '0' is right to left and '90' is top to bottom, and '270' bottom to top
          
          [env: SWWW_TRANSITION_ANGLE=]
          [default: 45]

      --transition-pos <TRANSITION_POS>
          This is only used for the 'grow','outer' transitions. It controls the center of circle (default is 'center').
          
          Position values can be given in both percentage values and pixel values: float values are interpreted as percentages and integer values as pixel values eg: 0.5,0.5 means 50% of
          the screen width and 50% of the screen height 200,400 means 200 pixels from the left and 400 pixels from the bottom
          
          the value can also be an alias which will set the position accordingly): 'center' | 'top' | 'left' | 'right' | 'bottom' | 'top-left' | 'top-right' | 'bottom-left' |
          'bottom-right'
          
          [env: SWWW_TRANSITION_POS=]
          [default: center]

      --invert-y
          inverts the y position sent in 'transiiton_pos' flag
          
          [env: INVERT_Y=]

      --transition-bezier <TRANSITION_BEZIER>
          bezier curve to use for the transition https://cubic-bezier.com is a good website to get these values from
          
          eg: 0.0,0.0,1.0,1.0 for linear animation
          
          [env: SWWW_TRANSITION_BEZIER=]
          [default: .54,0,.34,.99]

      --transition-wave <TRANSITION_WAVE>
          currently only used for 'wave' transition to control the width and height of each wave
          
          [env: SWWW_TRANSITION_WAVE=]
          [default: 20,20]

  -h, --help
          Print help (see a summary with '-h')


fabric-nix/my_project/wallpaper/settings.py
from fabric import Application
from fabric.widgets.wayland import WaylandWindow as Window
from combobox import ComboBox
from fabric.widgets.label import Label
from fabric.widgets.box import Box
from fabric.widgets.scale import Scale, ScaleMark
from fabric.widgets.scrolledwindow import ScrolledWindow

import textwrap
import gi
gi.require_version("Gtk", "3.0")
from gi.repository import Gtk

from separator import Separator
from fabric.utils import invoke_repeater, get_relative_path



# TODO add tooltip and get it from the help, just copy paste => Description
# TODO disable/enable optins in function of value of animation
class WallpaperSettings(Window):
    def __init__(self, **kwargs):
        super().__init__(
            #layer="top",
            anchor="left top right",
            #anchor="left bottom top right",
            #exclusivity="auto",
            **kwargs
        )
        # TODO add random stuff for all value when possible
        # TODO add submit button and store value in a text
        scale_1 = Scale(
            value=90, min_value=0, max_value=255,
        )
        scale_2 = Scale(
            value=10, min_value=0, max_value=300,
        )
        box = Box(
            name="outer-box",
            open_inspector=True,
            orientation="vertical",
            children=[
                Separator(orientation="horizontal"),
                Label(label="SWWW Settings").build().add_style_class("title").unwrap(),
                Separator(orientation="horizontal"),

                  #     --fill-color <FILL_COLOR>
  #         Which color to fill the padding with when output image does not fill screen
          
  #         [default: 000000]

                #.build().add_style_class("img").unwrap()
                # Resize
                Box(
                    h_align="center",
                    orientation="vertical",
                    children=[
                        Separator(orientation="horizontal"),
                        Label(h_align="start", label="Resize").build().add_style_class("label-title").unwrap(),
                        # Label(h_align="start", label="""
                        # plpo
                        # """).build().add_style_class("label-descriptio").unwrap(),
                        Label(h_align="start", label=textwrap.dedent("""
                            - no:   Do not resize the image
                            - crop: Resize the image to fill the whole screen, cropping out parts that don't fit
                            - fit:  Resize the image to fit inside the screen, preserving the original aspect ratio
                        """)).build().add_style_class("label-description").unwrap(),
                        ComboBox(items=["no", "crop", "fit"]),

                        # TODO Fill color
                        Separator(orientation="horizontal"),
                        Label(label="Fill color"),

                        # Filter
                        Separator(orientation="horizontal"),
                        Label(label="Filter"),
                        ComboBox(items=["Nearest", "Bilinear", "CatmullRom", "Mitchell", "Lanczos3"]),

                        # Transition type
                        Separator(orientation="horizontal"),

                        Label(label="Transition type"),
                        ComboBox(items=["none", "simple", "fade", "left", "right", "top", "bottom", "wipe", "wave", "grow", "center", "any", "outer", "random"]),

                        # Transtion step
                        Separator(orientation="horizontal"),
                        Label(label="Transition step"),
                        scale_1,

                        # Transtion duration
                        Separator(orientation="horizontal"),
                        Label(label="Transition duration (in seconds)"),
                        scale_2,
                        Label(label=f"value is {scale_2.get_value()}"),

                        # Transition fps
                        Separator(orientation="horizontal"),
                        Label(label="Transition fps"),
                        Scale(value=90, min_value=0, max_value=1000),

                        # Transition angle
                        Separator(orientation="horizontal"),
                        Label(label="Transition angle"),
                        Scale(value=0, min_value=0, max_value=360),

                        # Transition pos
                        Separator(orientation="horizontal"),
                        Label(label="Transition pos in %"),
                        Scale(value=0, min_value=0, max_value=100),

                        # Invert Y
                        Separator(orientation="horizontal"),
                        Label(label="Invert the Y")
                ]),


                # TODO booelan
  #     --transition-bezier <TRANSITION_BEZIER>
  #         bezier curve to use for the transition https://cubic-bezier.com is a good website to get these values from
          
  #         eg: 0.0,0.0,1.0,1.0 for linear animation
          
  #         [env: SWWW_TRANSITION_BEZIER=]
  #         [default: .54,0,.34,.99]

  #     --transition-wave <TRANSITION_WAVE>
  #         currently only used for 'wave' transition to control the width and height of each wave
          
  #         [env: SWWW_TRANSITION_WAVE=]
  #         [default: 20,20]
            ],
        )
        self.add(box)


if __name__ == "__main__":
    settings = WallpaperSettings()
    app = Application("wallpaper-settings", settings)
    app.set_stylesheet_from_file(get_relative_path("./style.css"))
    app.run()
